src/app-nordigen/services/nordigen-service.js

```diff
@@ -50,10 +51,44 @@ export const nordigenService = {
    * @returns {Promise<void>}
    */
   setToken: async () => {
-    if (!nordigenClient.token) {
+    const isExpiredJwtToken = (token) => {
+      const decodedToken = jwt.decode(token);
+      if (!decodedToken) {
+        return true;
+      }
+      const payload = decodedToken.payload;
+      const clockTimestamp = Math.floor(Date.now() / 1000);
+      return clockTimestamp >= payload.exp;
+    };
+
+    const accessToken = nordigenClient.token;
+    const refreshToken = process.env.NORDIGEN_REFRESH_TOKEN;
+    const accessTokenExpired = isExpiredJwtToken(accessToken);
+    const refreshTokenExpired = isExpiredJwtToken(refreshToken);
+
+    if (!accessTokenExpired) {
+      return;
+    }
+
+    if (!refreshTokenExpired) {
+      const newToken = await client.exchangeToken({
+        refreshToken: refreshToken,
+      });
+
+      nordigenClient.token = newToken.access;
+      process.env.NORDIGEN_REFRESH_TOKEN = undefined;
+    } else {
+      // Generate new access token. Token is valid for 24 hours
       const tokenData = await client.generateToken();
+
+      console.log({ msg: 'Generated new token', tokenData });
+      // Get access and refresh token
+      // Note: access_token is automatically injected to other requests after you successfully obtain it
+      const token = tokenData.access;
+      process.env.NORDIGEN_REFRESH_TOKEN = tokenData.refresh;
```

Comment:
What happens if I restart my server? That would clear the `NORDIGEN_REFRESH_TOKEN` env var, right? Which means I need to re-authenticate again every linked account? Is my assumption here correct?

Reply:
Basically my point is that we'll want to store this refresh token in the database so it is persisted between sessions.

Reply:
No, this token is only used to authenticate the Actual app to communicate with Nordigen API. Your agreement/requisition created after linking accounts still allows authenticated users to read data from your banks.

The first generated token is valid 24h after creation. Together with them, we receive a refresh token which is valid for 30 days. Thanks to the refresh token we can call for a refresh token but after this call, we receive only a fresh token valid for 24h but we don't receive a new refresh token.  

Example: 
- Day 1 -> Run sync transfers - the app will generate an `access token` and `refresh token`
- Day 2 -> Run sync transfers - the `access token` expired so we generate a new token using the `refresh token`
- Day 3 -> Run sync transfers - The `access token` generated on Day 2 expired, we don't have a `refresh token` so we have to generate new `access token` and `refresh token` similar like in Day 1

Reply:
If we can generate a new `access token` without a `refresh token` (day 3): why do we even need to implement logic for `refresh token`? Why not generate a new `access token` whenever it expires (day 1)?

Reply:
Obviously, we can ;) I don't see any strong pros to using a `refresh token`. I will change it to use only the `access token` when it will expired without playing with the `refresh token`

Reply:
IMO using just a single token will make the code much easier to read. So if it's not vital to use the `refresh token`, then I'd recommend not to use it.

(I tried looking at Nordigen docs to understand why they give a refresh token, but I could not find my answer.. they give answers about "how", but not the "why")

Reply:
I had the same issue when I tried to understand it

Reply:
I updated the code